### ISSUE TITLE
Auto-fix for Validators

### ISSUE THREAD

**Issue by tokejepsen** [pyblish/pyblish-base#43]
- Function to execute "fix" function on validators
- Way to find out whether or not a validator has auto-fix code
- Validators "fix" function needs to return errors

**Comment by mottosso** [comment]
> Way to find out whether or not a validator has auto-fix code

The duck in me thinks that the way this should get discovered is by looking for the existence of the `fix` method on validator instances.

``` python
# E.g.
>>> if hasattr(validator, 'fix'):
...     # Auto fixing!
...     validator.fix()
```

Or in the UI:

``` python
>>> if hasattr(validator, 'fix'):
...     ui.fix_button.show()
```

The problem though is that we wouldn't be able to have this method be available in the abstract base-class; as it would make it available in all classes, even those without implementation. And by not having it there, we would have to rely on another mechanism for documenting it.

> Validators "fix" function needs to return errors

The Validators themselves raises errors, and passes them along to the relevant instance. Maybe something similar could work for auto-fixes. They could start by clearing out existing exceptions from its instance.

**Comment by tokejepsen** [comment]
> The problem though is that we wouldn't be able to have this method be available in the abstract base-class; as it would make it available in all classes, even those without implementation. And by not having it there, we would have to rely on another mechanism for documenting it.

Was thinking the same. I wonder about having an auto-fix attribute on the validator, so we aren't replying on the fix function being there or not.

``` python
>>> if validator.auto_fix:
... validator.fix()
```

Even better would be to have some function on the validators, that checks the fix function for code and return True/False.

> The Validators themselves raises errors, and passes them along to the relevant instance. Maybe something similar could work for auto-fixes. They could start by clearing out existing exceptions from its instance.

Using the existing error handling, and flushing them seems like a good idea:)

**Comment by mottosso** [comment]
> Using the existing error handling, and flushing them seems like a good idea:)

On second thought though, I wouldn't expect a fix to automatically run another validation, but rather clear out the "green light" on that validator so the user could run it himself.

In that case, it may be better if the validation started out by flushing existing exceptions.

What do you think?

**Comment by tokejepsen** [comment]
> On second thought though, I wouldn't expect a fix to automatically run another validation, but rather clear out the "green light" on that validator so the user could run it himself.

Not entire sure what you mean? I wouldn't have a validator trigger another validator, which makes for less composable code. Workflow:
- flush existing exceptions
- run auto-fix code on context
- raise exceptions related to auto-fixing
  - this should most of the time include the exceptions from the process function

> In that case, it may be better if the validation started out by flushing existing exceptions.

Think we are on the same page:)

**Comment by BigRoy** [comment]
Probably a bit too much of an abstraction, but I'm also thinking about how to hook up with a UI.

What if each Filter had a list of actions they provide (which can be appended to per subclass), eg. a dict:

{'delete': self.delete, 'fix': self.fix}

This way we can also easily vary how the UI differs for a Validator and Selector.
And we can set-up the actions workflow on the base class.

**Comment by tokejepsen** [comment]
> What if each Filter had a list of actions they provide (which can be appended to per subclass), eg. a dict:

Think I know what you mean:

``` python
class Filter(object):
    actions = dict()

class myValidator(Filter):
   actions['fix'] = self.fix
   def fix(self):
      #some code

val = myValidator()
if hasattr(val.actions, 'fix'):
    val.fix()
```

Dunno if my syntax is entirely correct, but hope you get the point.

**Comment by BigRoy** [comment]
Yeah, something like that. Though I don't think you can do hasattr() on a dict. You would do `'fix' in actions` check or actions.get('fix', None). Though that's just syntax, the overall idea is right.

And the UI could show the buttons by just showing what's in the dict (iterating over it in a sorted manner?)

**Comment by mottosso** [comment]
> Not entire sure what you mean? I wouldn't have a validator trigger another validator, which makes for less composable code.

Something like this:
1. The user runs a validation and the validation fails. 
2. When it fails, it could give the user an option to fix. 
3. When the user hits "fix" it _only_ fixes. 
4. From there, the user could run another validation, if he wanted to.

The alternative, is for another validation to run immediately after "fix" has been pressed. This is what I wouldn't expect.

This illustration might help clarify what I mean.
https://github.com/abstractfactory/publish/wiki/Process#process

**Comment by tokejepsen** [comment]
@mottosso, cool:)

> Yeah, something like that. Though I don't think you can do hasattr() on a dict. You would do `'fix' in actions` check or actions.get('fix', None). Though that's just syntax, the overall idea is right.

Think this is a cool idea. So the only function that gets set from the begining is the "process", all other functions are add with the actions dict?

> This way we can also easily vary how the UI differs for a Validator and Selector.
> And we can set-up the actions workflow on the base class.

Are you saying that Selectors should have Filter as a base class?
Had problems with the initialization, as it required an instance, but maybe you have an idea to work around that?

**Comment by BigRoy** [comment]
I would personally inherit from same Base class, though I read about your issues. The design I had in mind at first was more like a node graph where each node (=filter) acted only on _attributes_ within a _context_ and that methodology would be used for all: Selectors, Validator and Extractors.

In this way each Filter is a black box only operating on requested data (inputs) and can only alter data that's been defined as  node output.

I am meeting up with Marcus in Amsterdam tonight and we'll probably discuss some design decisions more closely. Will write a recap tomorrow if there are some new ideas. ;)

**Comment by mottosso** [comment]
This was implemented in #95
